resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1416,4 +1389,136 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+         */
+        void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture);
+
+        /**
+         * Close the {@link DnsResolveChannelProvider} and so cleanup resources if needed.
+         */
+        void close();
+    }
+
+    private static final class DnsResolveSameChannelProvider implements DnsResolveChannelProvider {
+
+        private final ChannelFuture resolveChannelFuture;
+
+        DnsResolveSameChannelProvider(Bootstrap bootstrap, SocketAddress localAddress) {
+            SocketAddress address = localAddress == null ? new InetSocketAddress(0) : localAddress;
+            resolveChannelFuture = bootstrap.bind(address);
+        }
+
+        @Override
+        public ChannelFuture nextResolveChannel() {
+            return resolveChannelFuture;
+        }
+
+        @Override
+        public void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture) {
+            // Do nothing
+        }
+
+        @Override
+        public void close() {
+            resolveChannelFuture.channel().close();
+        }
+    }
+
+    private static final class DnsResolveSameChannelUntilTimeoutProvider implements DnsResolveChannelProvider {
+        private final Bootstrap bootstrap;
+        private final SocketAddress localAddress;
+        private volatile ChannelFuture resolveChannelFuture;
+
+        DnsResolveSameChannelUntilTimeoutProvider(Bootstrap bootstrap, SocketAddress localAddress) {
+            this.bootstrap = bootstrap;
+            this.localAddress = localAddress;
+        }
+
+        @Override
+        public ChannelFuture nextResolveChannel() {
+            if (resolveChannelFuture == null) {
+                SocketAddress address = localAddress == null ? new InetSocketAddress(0) : localAddress;
+                resolveChannelFuture = bootstrap.bind(address);
+            }
+            return resolveChannelFuture;
+        }
+
+        @Override
+        public void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture) {
+            if (isTimeoutError(resolveFuture.cause())) {
+                this.resolveChannelFuture = null;
+                // Let's close the Channel now... This might also cause a few other failures for in-progress queries
+                // We could also consider to count how many outstanding resolves we have and only after we processed
+                // all of them close the Channel.
+                resolveChannelFuture.channel().close();
+            }
+        }
+
+        @Override
+        public void close() {
+            ChannelFuture resolveChannelFuture = this.resolveChannelFuture;
+            if (resolveChannelFuture != null) {
+                resolveChannelFuture.channel().close();
+            }
+        }
+    }
+
+    private static final class DnsResolveNewChannelProvider implements DnsResolveChannelProvider {
```

Comment:
This is the only one that is used atm, but we could make it configurable to use one of the three here.

Reply:
We definitely need a config. It would be safer to preserve existing behavior for the first release (`DnsResolveSameChannelProvider`) and use other providers only when configured. Either via a builder or a system property.

Reply:
Yeah I agree … just wanted to show something first